### ISSUE TITLE
Student Performance Prediction Using Machine Learning

### ISSUE THREAD

**Issue by singipurapudinesh-coder** [singipurapudinesh-coder/DATA-SCIENCE-INTERNSHIP-PROJECTS#2]
<html>
<body>
<html><head></head><body><h1>🎓 Student Performance Prediction Using Machine Learning</h1><blockquote><p><strong>A Data Science and Machine Learning project that analyzes student academic and behavioral factors and predicts final student performance using regression algorithms.</strong></p></blockquote><p><a href="https://www.python.org/"><img src="https://img.shields.io/badge/Python-3.x-blue?logo=python" alt="Python"></a><br><a href="https://pandas.pydata.org/"><img src="https://img.shields.io/badge/Pandas-Data%20Analysis-150458?logo=pandas" alt="Pandas"></a><br><a href="https://scikit-learn.org/"><img src="https://img.shields.io/badge/Scikit--Learn-Machine%20Learning-F7931E?logo=scikit-learn" alt="Scikit-Learn"></a><br><a href="https://matplotlib.org/"><img src="https://img.shields.io/badge/Matplotlib-Visualization-11557C" alt="Matplotlib"></a><br><a href="https://jupyter.org/"><img src="https://img.shields.io/badge/Jupyter-Notebook-F37626?logo=jupyter" alt="Jupyter"></a></p><hr><h2>📌 Overview</h2><p><strong>Student Performance Prediction</strong> is an end-to-end Machine Learning project developed to analyze student academic, attendance, study, and lifestyle-related factors and predict their final academic score.</p><p>The project follows a complete Data Science workflow:</p><pre><code class="language-text">Dataset
   ↓
Data Cleaning &amp; Preprocessing
   ↓
Exploratory Data Analysis
   ↓
Feature Engineering
   ↓
Train/Test Split
   ↓
Model Training
   ↓
Model Evaluation
   ↓
Best Model Selection
   ↓
New Student Prediction
</code></pre><p>The system compares <strong>Linear Regression</strong> and <strong>Random Forest Regression</strong>, evaluates their performance using multiple regression metrics, automatically selects the best model, and uses the trained model to predict the performance of new students.</p><hr><h2>🎯 Problem Statement</h2><p>Educational institutions collect large amounts of student data, but analyzing this information manually can make it difficult to identify performance patterns and predict future academic outcomes.</p><p>This project addresses the problem by developing a Machine Learning system capable of:</p><ul><li><p>Analyzing student performance data</p></li><li><p>Identifying relationships between different student factors</p></li><li><p>Training regression models</p></li><li><p>Comparing model performance</p></li><li><p>Predicting final student scores</p></li><li><p>Categorizing predicted student performance</p></li></ul><hr><h2>🚀 Key Features</h2><ul><li><p>📊 Exploratory Data Analysis</p></li><li><p>🧹 Data cleaning and preprocessing</p></li><li><p>🔎 Feature analysis and correlation analysis</p></li><li><p>📈 Multiple data visualizations</p></li><li><p>🤖 Linear Regression model</p></li><li><p>🌲 Random Forest Regression model</p></li><li><p>📏 MAE, MSE, RMSE and R² evaluation</p></li><li><p>🏆 Automatic best-model selection</p></li><li><p>💾 Trained model persistence using Pickle</p></li><li><p>🎓 New student performance prediction</p></li><li><p>📉 Actual vs Predicted visualization</p></li><li><p>📊 Model comparison visualization</p></li><li><p>🖥️ Command-line project interface</p></li><li><p>📓 Jupyter Notebook analysis</p></li><li><p>📁 Organized project architecture</p></li></ul><hr><h2>🛠️ Tech Stack</h2>
Technology | Purpose
-- | --
Python | Core programming language
Pandas | Data manipulation and analysis
NumPy | Numerical computation
Scikit-learn | Machine Learning and evaluation
Matplotlib | Data visualization
Seaborn | Statistical visualization
Jupyter Notebook | Interactive analysis
Pickle | Model serialization

<h3>🥇 Selected Model: Linear Regression</h3><p><strong>R² Score: 0.8681</strong></p><p>The Linear Regression model achieved the highest R² score among the evaluated models and was therefore selected as the final prediction model.</p><p>An R² of approximately <strong>0.868</strong> indicates that the model explains about <strong>86.8% of the variation in final student scores</strong> on the test set.</p><hr><h2>🔮 New Student Prediction</h2><p>The project includes a prediction module that accepts information for a new student and generates an estimated final score.</p><h3>Example Input</h3><pre><code class="language-text">Gender: Female
Age: 18
Study Hours: 5
Attendance: 90
Previous Score: 78
Assignment Score: 85
Midterm Score: 82
Extracurricular: Yes
Internet Access: Yes
Sleep Hours: 7
</code></pre><h3>Example Output</h3><pre><code class="language-text">============================================================
STUDENT PERFORMANCE PREDICTION RESULT
============================================================

Predicted Final Score: XX.XX / 100
Performance Category: Very Good

============================================================
</code></pre><p>The prediction value is generated dynamically by the trained Machine Learning model.</p><hr><h2>📂 Project Structure</h2><pre><code class="language-text">Student-Performance-Prediction/
│
├── data/
│   └── student_performance.csv
│
├── notebooks/
│   └── student_performance_analysis.ipynb
│
├── src/
│   ├── __init__.py
│   ├── data_preprocessing.py
│   ├── eda.py
│   ├── model_training.py
│   └── prediction.py
│
├── models/
│   └── student_performance_model.pkl
│
├── outputs/
│   ├── graphs/
│   │   ├── final_score_distribution.png
│   │   ├── study_hours_vs_score.png
│   │   ├── attendance_vs_score.png
│   │   ├── previous_score_vs_final_score.png
│   │   ├── correlation_heatmap.png
│   │   ├── actual_vs_predicted.png
│   │   └── model_comparison.png
│   │
│   └── predictions/
│       └── model_results.csv
│
├── main.py
├── requirements.txt
├── README.md
└── .gitignore
</code></pre><hr><h2>⚙️ Installation</h2><h3>1. Clone the Repository</h3><pre><code class="language-bash">git clone &lt;YOUR-GITHUB-REPOSITORY-URL&gt;
</code></pre><h3>2. Navigate to the Project</h3><pre><code class="language-bash">cd Student-Performance-Prediction
</code></pre><h3>3. Create a Virtual Environment</h3><pre><code class="language-bash">python -m venv venv
</code></pre><h3>4. Activate the Environment</h3><p><strong>Windows:</strong></p><pre><code class="language-powershell">venv\Scripts\activate
</code></pre><p><strong>Linux/macOS:</strong></p><pre><code class="language-bash">source venv/bin/activate
</code></pre><h3>5. Install Dependencies</h3><pre><code class="language-bash">pip install -r requirements.txt
</code></pre><hr><h2>▶️ Running the Project</h2><p>The project can be launched through the main menu:</p><pre><code class="language-bash">python main.py
</code></pre><p>You will see:</p><pre><code class="language-text">======================================================================
       STUDENT PERFORMANCE PREDICTION SYSTEM
======================================================================

Select an option:

1. Run Exploratory Data Analysis (EDA)
2. Train Machine Learning Models
3. Predict Student Performance
4. Run Complete Project
5. Exit
</code></pre><h3>Run EDA</h3><pre><code class="language-text">1
</code></pre><h3>Train Models</h3><pre><code class="language-text">2
</code></pre><h3>Predict a New Student</h3><pre><code class="language-text">3
</code></pre><h3>Run the Complete Pipeline</h3><pre><code class="language-text">4
</code></pre><hr><h2>📓 Jupyter Notebook</h2><p>The complete analytical workflow is available in:</p><pre><code class="language-text">notebooks/student_performance_analysis.ipynb
</code></pre><p>The notebook covers:</p><ul><li><p>Dataset loading</p></li><li><p>Dataset exploration</p></li><li><p>Data cleaning</p></li><li><p>Statistical analysis</p></li><li><p>EDA</p></li><li><p>Visualization</p></li><li><p>Feature preparation</p></li><li><p>Model training</p></li><li><p>Model evaluation</p></li><li><p>Prediction</p></li></ul><p>Launch Jupyter with:</p><pre><code class="language-bash">jupyter notebook
</code></pre><hr><h2>📸 Project Outputs</h2><p>The project generates visual outputs under:</p><pre><code class="language-text">outputs/graphs/
</code></pre><p>including:</p><pre><code class="language-text">final_score_distribution.png
study_hours_vs_score.png
attendance_vs_score.png
previous_score_vs_final_score.png
correlation_heatmap.png
actual_vs_predicted.png
model_comparison.png
</code></pre><p>Model evaluation results are stored in:</p><pre><code class="language-text">outputs/predictions/model_results.csv
</code></pre><p>The trained model is stored as:</p><pre><code class="language-text">models/student_performance_model.pkl
</code></pre><hr><h2>🧠 Project Architecture</h2><pre><code class="language-text">                    ┌─────────────────────┐
                    │ Student Dataset     │
                    │ CSV File            │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ Data Preprocessing  │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ Exploratory         │
                    │ Data Analysis       │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ Train/Test Split    │
                    └──────────┬──────────┘
                               │
                 ┌─────────────┴─────────────┐
                 ▼                           ▼
       ┌──────────────────┐        ┌──────────────────┐
       │ Linear Regression│        │ Random Forest    │
       └────────┬─────────┘        └────────┬─────────┘
                │                           │
                └─────────────┬─────────────┘
                              ▼
                    ┌─────────────────────┐
                    │ Model Evaluation    │
                    │ MAE / MSE / RMSE    │
                    │ R² Score            │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ Best Model          │
                    │ Linear Regression    │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ New Student Data    │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │ Predicted Score     │
                    └─────────────────────┘
🎓 Academic Context</h2><p>This project was developed as part of the:</p><p><strong>Data Science Major Project — Topgrade Innovation</strong></p><h3>Project Category</h3><pre><code class="language-text">Data Science
Machine Learning
Predictive Analytics
</code></pre><h3>Problem Type</h3><pre><code class="language-text">Supervised Learning
Regression
</code></pre><h3>Target</h3><pre><code class="language-text">Student Final Score
</code></pre><h3>Selected Algorithm</h3><pre><code class="language-text">Linear Regression
</code></pre><h3>Best R² Score</h3><pre><code class="language-text">0.8681
</code></pre><hr><h2>👨‍💻 Author</h2><h3>Singipurapu Dinesh</h3><p><strong>Computer Science Student | Python Developer | AI &amp; Data Science Enthusiast</strong></p><p>Interested in building practical solutions using:</p><ul><li><p>Python</p></li><li><p>Data Science</p></li><li><p>Machine Learning</p></li><li><p>Artificial Intelligence</p></li><li><p>Data Analytics</p></li></ul><hr><h2>⭐ Support</h2><p>If you find this project useful, consider giving the repository a ⭐ on GitHub.
</body>
</html>
